### PR TITLE
Do not return downlink token if the downlink is nil

### PR DESCRIPTION
**Pull request by KrishnaIyer**
#### Summary

Do not return downlink token if the downlink is nil. Fixes the panic reported on Slack.

#### Changes

- If the `downlink` message is nil, return false.

#### Testing

I think if CI passes this is good.

##### Regressions

Don't see any.

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Review by johanstokking — COMMENTED** [review]
This is `nil` in and `nil` out. I don't know why `DownlinkTokens` would support putting `nil` messages but not returning them. How do they end up in the downlink tokens in the first place? Why isn't the caller of `Get()` allowing for a `nil` result?

Note that downlink tokens are not only used to store downlink messages, but also for round-trip times. So `nil, DURATION, true` is a useful result of `Get()`.

**Comment by KrishnaIyer** [comment]
> Note that downlink tokens are not only used to store downlink messages, but also for round-trip times. So nil, DURATION, true is a useful result of Get().

Hmm? Where is a situation where downlink (that is supposedly sent to the gateway) is nil but there's an entry for `downlinkToken.time`? How do we send a nil downlink and then check the corresponding Tx ACK for RTT?

> How do they end up in the downlink tokens in the first place?

Because when the key is 0 and the slice is empty, we just return an empty `downlinkToken` whose `msg` is nil.
https://github.com/TheThingsNetwork/lorawan-stack/blob/19f5503f16c986f1c71c7b72f38f4a0a677a591e/pkg/gatewayserver/io/downlink_token.go#L59

We could indeed just check for nil value on the caller, but I don't see any valid case for `nil, DURATION, true`.

What's worse, since we're using the default value `time.Time` in the RTT, we would be skewing the RTT value.

File: pkg/gatewayserver/io/udp/udp.go [hunk 1]
```diff
@@ -328,7 +328,8 @@ func (s *srv) handleUp(ctx context.Context, state *state, packet encoding.Packet
 			}
 		}
 		var rtt *time.Duration
-		if downlink, delta, ok := state.tokens.Get(uint16(packet.Token[0])<<8|uint16(packet.Token[1]), packet.ReceivedAt); ok {
+		downlink, delta, ok := state.tokens.Get(uint16(packet.Token[0])<<8|uint16(packet.Token[1]), packet.ReceivedAt)
+		if ok && downlink != nil {
 			msg.TxAcknowledgment.DownlinkMessage = downlink
 			msg.TxAcknowledgment.CorrelationIDs = downlink.CorrelationIDs
 			rtt = &delta
```
Comment (KrishnaIyer) [inline pkg/gatewayserver/io/udp/udp.go]: I don't see a case where there's no downlink but a corresponding time stamp is stored to check the TxAck against.

**Comment by adriansmares** [comment]
@johanstokking keep in mind that your assumption is that the value _was_ put there via `Next`. This is not always the case.

If `uint16(packet.Token[0])<<8|uint16(packet.Token[1])` is a multiple of `downlinkTokenItems` (or simply 0), and the connection is new (maybe due to a reconnect, like the spurious ones we're currently encountering) the following condition will be false:
https://github.com/TheThingsNetwork/lorawan-stack/blob/3bc2c96246a7421fe7bba320f2e0a378a6e400c5/pkg/gatewayserver/io/downlink_token.go#L60
But there is no item there - it's just the zero value of `downlinkToken`. Both `down` and `time` will be `nil`/`0`. 

In my opinion 
https://github.com/TheThingsNetwork/lorawan-stack/blob/3bc2c96246a7421fe7bba320f2e0a378a6e400c5/pkg/gatewayserver/io/downlink_token.go#L60
is too weak, because I don't get `something, something, false` if I call `Get` with a token of `0` on a newly initialized `DownlinkTokens`.

**Review by johanstokking — COMMENTED** [review]
Apologies, please revert the original diff. There is no `nil` in `nil` out; an uninitialized `downlinkTokens` at position `0` (initial state) results in this behavior and this should be guarded for by `DownlinkTokens`.

File: pkg/gatewayserver/io/downlink_token.go [hunk 2]
```diff
@@ -57,7 +57,7 @@ func (t *DownlinkTokens) Next(msg *ttnpb.DownlinkMessage, time time.Time) uint16
 func (t DownlinkTokens) Get(token uint16, time time.Time) (*ttnpb.DownlinkMessage, time.Duration, bool) {
 	pos := token % downlinkTokenItems
 	item := t.items[pos]
-	if item.key != token {
+	if item.key != token || item.msg == nil {
```
Comment (adriansmares) [inline pkg/gatewayserver/io/downlink_token.go]: This is fine, but you may also check if the time is 0.

**Review by adriansmares — APPROVED** [review]

File: pkg/gatewayserver/io/downlink_token.go (hunk 2 above)
Comment (KrishnaIyer) [reply pkg/gatewayserver/io/downlink_token.go]: Indeed. I don't see a case where that should be possible but better to do that here.

**Pull request merged**